source/OctopusTools/Commands/CleanEnvironmentCommand.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using log4net;
+using Octopus.Client.Model;
+using OctopusTools.Infrastructure;
+using OctopusTools.Util;
+
+namespace OctopusTools.Commands
+{
+    [Command("clean-environment", Description = "Cleans all Offline Machines from an Environment")]
+    public class CleanEnvironmentCommand : ApiCommand
+    {
+        string environmentName;
+        MachineModelStatus status = MachineModelStatus.Unknown;
+
+        public CleanEnvironmentCommand(IOctopusRepositoryFactory repositoryFactory, ILog log, IOctopusFileSystem fileSystem)
+            : base(repositoryFactory, log, fileSystem)
+        {
+            var options = Options.For("Cleanup");
+            options.Add("environment=", "Name of an environment to clean up.", v => environmentName = v);
+            options.Add("status=", "Status of Machines to clean up (Online, Offline, NeedsUpgrade, CalamariNeedsUpgrade, Disabled).", v =>
+            {
+                Enum.TryParse(v, true, out status);
+            });
+        }
+
+        protected override void Execute()
+        {
+            if (string.IsNullOrWhiteSpace(environmentName)) throw new CommandException("Please specify an environment name using the parameter: --environment=XYZ");
+            if (status == MachineModelStatus.Unknown) throw new CommandException("Please specify status using the parameter: --status=Offline");
+
+            Log.Debug("Loading environment...");
+            var environmentResource = Repository.Environments.FindByName(environmentName);
+
+            Log.Debug("Loading machines...");
+            var machines = Repository.Machines.FindMany(x =>
+            {
+                return x.Status == status && x.EnvironmentIds.Any(environmentId => environmentId == environmentResource.Id);
+            });
+
+            if (machines != null)
+            {
+                Log.Info("Machines: " + machines.Count);
+
+                foreach (var machine in machines)
+                {
+                    Log.InfoFormat("Deleting - {0} {1} (ID: {2})", machine.Name, machine.Status, machine.Id);
+                    Repository.Machines.Delete(machine);
```

Comment:
Could a machine be a different status in each environment though?  Maybe my
selection logic is too simplistic.

On 27 January 2016 at 11:05, Paul Stovell <notifications@github.com> wrote:

> In source/OctopusTools/Commands/CleanEnvironmentCommand.cs
> <https://github.com/OctopusDeploy/Octopus-Tools/pull/69#discussion_r50928232>
> :
>
> > +            var environmentResource = Repository.Environments.FindByName(environmentName);
> > +
> > +            Log.Debug("Loading machines...");
> > +            var machines = Repository.Machines.FindMany(x =>
> > +            {
> > +                return x.Status == status && x.EnvironmentIds.Any(environmentId => environmentId == environmentResource.Id);
> > +            });
> > +
> > +            if (machines != null)
> > +            {
> > +                Log.Info("Machines: " + machines.Count);
> > +
> > +                foreach (var machine in machines)
> > +                {
> > +                    Log.InfoFormat("Deleting - {0} {1} (ID: {2})", machine.Name, machine.Status, machine.Id);
> > +                    Repository.Machines.Delete(machine);
>
> A machine may belong to more than one environment. We should check if that
> is the case and just remove the machine from the environment instead of
> deleting it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OctopusDeploy/Octopus-Tools/pull/69/files#r50928232>.
>



--